pkg/deploy/tls.go

```diff
@@ -0,0 +1,448 @@
+			return reconcile.Result{RequeueAfter: time.Second}, err
+		}
+		// Recreate the secret
+		return reconcile.Result{RequeueAfter: time.Second}, err
+	}
+
+	// Check owner reference
+	if cheTLSSecret.ObjectMeta.OwnerReferences == nil {
+		// Set owner Che cluster as Che TLS secret owner
+		if err := controllerutil.SetControllerReference(checluster, cheTLSSecret, clusterAPI.Scheme); err != nil {
+			logrus.Errorf("Failed to set owner for Che TLS secret \"%s\". Error: %s", cheTLSSecretName, err)
+			return reconcile.Result{RequeueAfter: time.Second}, err
+		}
+		if err := clusterAPI.Client.Update(context.TODO(), cheTLSSecret); err != nil {
+			logrus.Errorf("Failed to update owner for Che TLS secret \"%s\". Error: %s", cheTLSSecretName, err)
+			return reconcile.Result{RequeueAfter: time.Second}, err
+		}
+	}
+
+	// ===== Check Che CA certificate ===== //
+
+	cheTLSSelfSignedCertificateSecret := &corev1.Secret{}
+	err = clusterAPI.Client.Get(context.TODO(), types.NamespacedName{Namespace: checluster.Namespace, Name: CheTLSSelfSignedCertificateSecretName}, cheTLSSelfSignedCertificateSecret)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			// Error reading Che self-signed secret info
+			logrus.Errorf("Error getting Che self-signed certificate secert \"%s\": %v", CheTLSSelfSignedCertificateSecretName, err)
+			return reconcile.Result{RequeueAfter: time.Second}, err
+		}
+		// Che CA self-signed cetificate secret doesn't exist.
+		// This means that commonly trusted certificate is used.
+	} else {
+		// Che CA self-signed certificate secret exists, check for required data fields
+		if !isCheCASecretValid(cheTLSSelfSignedCertificateSecret) {
+			logrus.Infof("Che self-signed certificate secret \"%s\" is invalid. Recrating...", CheTLSSelfSignedCertificateSecretName)
+			// Che CA self-signed certificate secret is invalid, delete it
+			if err = clusterAPI.Client.Delete(context.TODO(), cheTLSSelfSignedCertificateSecret); err != nil {
+				logrus.Errorf("Error deleting Che self-signed certificate secret \"%s\": %v", CheTLSSelfSignedCertificateSecretName, err)
+				return reconcile.Result{RequeueAfter: time.Second}, err
+			}
+			// Also delete Che TLS as the certificates should be created together
+			// Here it is not mandatory to check Che TLS secret existence as it is handled above
+			if err = clusterAPI.Client.Delete(context.TODO(), cheTLSSecret); err != nil {
+				logrus.Errorf("Error deleting Che TLS secret \"%s\": %v", cheTLSSecretName, err)
+				return reconcile.Result{RequeueAfter: time.Second}, err
+			}
+			// Regenerate Che TLS certicates and recreate secrets
+			return reconcile.Result{RequeueAfter: time.Second}, nil
+		}
+
+		// Check owner reference
+		if cheTLSSelfSignedCertificateSecret.ObjectMeta.OwnerReferences == nil {
+			// Set owner Che cluster as Che TLS secret owner
+			if err := controllerutil.SetControllerReference(checluster, cheTLSSelfSignedCertificateSecret, clusterAPI.Scheme); err != nil {
```

Comment:
Is it better to move this code to `getSpecSecret`
We set controllerRefence for all components at getSpec func

Reply:
The reason why we add controller reference here is that the secret could be created by self-signed secrets generator job which doesn't set any reference.